rclcpp/include/rclcpp/topic_statistics/subscriber_topic_statistics.hpp

```diff
@@ -0,0 +1,194 @@
+
+
+namespace rclcpp
+{
+namespace topic_statistics
+{
+
+using libstatistics_collector::collector::GenerateStatisticMessage;
+using metrics_statistics_msgs::msg::MetricsMessage;
+using libstatistics_collector::moving_average_statistics::StatisticData;
+
+/**
+ * Class used to collect, measure, and publish topic statistics data. Current statistics
+ * supported for subscribers are received message age and received message period.
+ *
+ * @tparam CallbackMessageT the subscribed message type
+ */
+template<typename CallbackMessageT>
+class SubscriberTopicStatistics
+{
+  using topic_stats_collector =
+    libstatistics_collector::topic_statistics_collector::TopicStatisticsCollector<
+    CallbackMessageT>;
+  using received_message_period =
+    libstatistics_collector::topic_statistics_collector::ReceivedMessagePeriodCollector<
+    CallbackMessageT>;
+
+public:
+  /// Construct a SubcriberTopicStatistics object.
+  /**
+   * This object wraps utilities, defined in libstatistics_collector, to collect,
+   * measure, and publish topic statistics data.
+   *
+   * @param node_name the name of the node, which created this instance, in order to denote
+   * topic source
+   * @param publisher instance constructed by the node in order to publish statistics data
+   */
+  SubscriberTopicStatistics(
+    const std::string & node_name,
+    const rclcpp::Publisher<metrics_statistics_msgs::msg::MetricsMessage>::SharedPtr & publisher)
+  : node_name_(node_name),
+    publisher_(std::move(publisher))
+  {
+    // TODO(dbbonnie): ros-tooling/aws-roadmap/issues/226, received message age
+
+    auto rmp = std::make_unique<received_message_period>();
+    rmp->Start();
+    subscriber_statistics_collectors_.emplace_back(std::move(rmp));
+
+    window_start_ = rclcpp::Time(GetCurrentNanosecondsSinceEpoch());
+  }
+
+  virtual ~SubscriberTopicStatistics()
+  {
+    TearDown();
+  }
+
+  /// Stop all collectors, clear measurements, stop publishing timer, and reset publisher.
+  virtual void TearDown()
+  {
+    for (auto & collector : subscriber_statistics_collectors_) {
+      collector->Stop();
+    }
+
+    subscriber_statistics_collectors_.clear();
+
+    if (publisher_timer_) {
+      publisher_timer_->cancel();
+      publisher_timer_.reset();
+    }
+    if (publisher_) {
+      publisher_.reset();
+    }
+  }
+
+  /// Set the timer used to publish statistics messages.
+  /**
+   * @param measurement_timer the timer to fire the publisher, created by the node
+   */
+  void SetPublisherTimer(const rclcpp::TimerBase::SharedPtr & publisher_timer)
+  {
+    publisher_timer_ = std::move(publisher_timer);
+  }
+
+  /// Handle a message received by the subscription to collect statistics.
+  /**
+   * @param received_message the message received by the subscription
+   * @param now_nanoseconds current time in nanoseconds
+   */
+  virtual void OnMessageReceived(
+    const CallbackMessageT & received_message,
```

Comment:
```suggestion
    const CallbackMessageT &,
```
You can then remove `(void) received_message`

Reply:
This doesn't look like it was ever done.